### PR TITLE
Decurrying

### PR DESCRIPTION
**Pull request by pvdrz**

### CODE REVIEW

File: alonzo-ty/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+use std::ops::Deref;
+
+pub trait BaseTy: Clone {
+    type Value: Clone;
+}
+
+#[derive(Clone)]
+pub enum Ty<T: BaseTy> {
+    Base(T),
+    Func(Box<Self>, Box<Self>),
```
Comment (zesterer) [inline alonzo-ty/src/lib.rs:10]: Hmm, I think we might want to specialize this type between `mir` and `lir`. For example, we probably want to include information about the function payload in `lir`.

Comment (zesterer) [reply alonzo-ty/src/lib.rs:10]: And, in general, we might be able to do more detailed lowering for other types that might warrant having definitions of `Ty` that are specific to `alonzo_lir` and `alonzo_hir`.

File: alonzo-ty/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    Product(Vec<Self>),
+    Sum(Vec<Self>),
+}
+
+#[derive(Clone)]
+pub struct TyNode<I, T: BaseTy> {
```
Comment (zesterer) [inline alonzo-ty/src/lib.rs:16]: I use a similar but more general pattern in Tao for things like this. Perhaps we can abstract it and use it for storing all kinds of metadata common to `alonzo-lir` and `alonzo-mir`? https://github.com/zesterer/tao/blob/master/src/node.rs

Comment (pvdrz) [reply alonzo-ty/src/lib.rs:16]: That sounds good.

File: alonzo-ty/src/lib.rs (hunk 1 above)
Comment (pvdrz) [reply alonzo-ty/src/lib.rs:10]: But should the function payload info be stored in the type of the object? What else could be different between these two?

Comment (zesterer) [reply alonzo-ty/src/lib.rs:10]: If we're lowering things from `mir` to `lir`, we will probably want to do other sort of lowering such as turning, say, lists of a known size into static arrays with the length as part of the type.

**Pull request closed without merging**